codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AccountIdEndpointModeUserAgent.java

```diff
@@ -0,0 +1,69 @@
+package software.amazon.smithy.aws.go.codegen.customization;
+
+import java.util.List;
+import java.util.Map;
+import software.amazon.smithy.aws.go.codegen.AwsGoDependency;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.GoCodegenContext;
+import software.amazon.smithy.go.codegen.GoDelegator;
+import software.amazon.smithy.go.codegen.GoSettings;
+import software.amazon.smithy.go.codegen.SmithyGoDependency;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+
+import static software.amazon.smithy.aws.go.codegen.customization.AccountIDEndpointRouting.hasAccountIdEndpoints;
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+import static software.amazon.smithy.go.codegen.SymbolUtils.buildPackageSymbol;
+
+/**
+ * Tracks the retry mode being used by the caller.
+ */
+public class AccountIdEndpointModeUserAgent implements GoIntegration {
+    private static final MiddlewareRegistrar MIDDLEWARE = MiddlewareRegistrar.builder()
+            .resolvedFunction(buildPackageSymbol("addUserAgentAccountIDEndpointMode"))
+            .useClientOptions()
+            .build();
+
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return List.of(
+                RuntimeClientPlugin.builder()
+                        .servicePredicate(AccountIDEndpointRouting::hasAccountIdEndpoints)
+                        .registerMiddleware(MIDDLEWARE)
+                        .build()
+        );
+    }
+
+    @Override
+    public void writeAdditionalFiles(GoCodegenContext ctx) {
+        if (!hasAccountIdEndpoints(ctx.model(), ctx.settings().getService(ctx.model()))) {
```

Comment:
do we need this if we have the service predicate?

Reply:
We could remove it but then it would just generate in every service and force a bunch of module updates.

Reply:
Ahh yeah, forgot that the predicate doesn't control whether we write something on the file or not. Yes, then this is perfect, let's not write superfluous code